include/ginkgo/core/reorder/rcm.hpp

```diff
@@ -0,0 +1,222 @@
+         */
+        bool GKO_FACTORY_PARAMETER_SCALAR(construct_inverse_permutation, false);
+
+        /**
+         * This parameter controls the strategy used to determine a starting
+         * vertex.
+         */
+        starting_strategy GKO_FACTORY_PARAMETER_SCALAR(
+            strategy, starting_strategy::pseudo_peripheral);
+    };
+    GKO_ENABLE_REORDERING_BASE_FACTORY(Rcm, parameters, Factory);
+    GKO_ENABLE_BUILD_METHOD(Factory);
+
+protected:
+    /**
+     * Generates the permutation matrix and if required the inverse permutation
+     * matrix.
+     */
+    void generate(std::shared_ptr<const Executor> &exec,
+                  std::unique_ptr<SparsityMatrix> adjacency_matrix) const;
+
+    explicit Rcm(std::shared_ptr<const Executor> exec)
+        : EnablePolymorphicObject<Rcm, ReorderingBase>(std::move(exec))
+    {}
+
+    explicit Rcm(const Factory *factory, const ReorderingBaseArgs &args)
+        : EnablePolymorphicObject<Rcm, ReorderingBase>(factory->get_executor()),
+          parameters_{factory->get_parameters()}
+    {
+        // Always execute the reordering on the cpu.
+        const auto is_gpu_executor =
+            this->get_executor() != this->get_executor()->get_master();
+        auto cpu_exec = is_gpu_executor ? this->get_executor()->get_master()
+                                        : this->get_executor();
+
+        auto adjacency_matrix = SparsityMatrix::create(cpu_exec);
+        Array<IndexType> degrees;
+
+        // The adjacency matrix has to be square.
+        GKO_ASSERT_IS_SQUARE_MATRIX(args.system_matrix);
+        // This is needed because it does not make sense to call the copy and
+        // convert if the existing matrix is empty.
+        if (args.system_matrix->get_size()) {
+            auto tmp = copy_and_convert_to<SparsityMatrix>(cpu_exec,
+                                                           args.system_matrix);
+            // This function provided within the Sparsity matrix format removes
+            // the diagonal elements and outputs an adjacency matrix.
+            adjacency_matrix = tmp->to_adjacency_matrix();
+        }
+
+        auto const dim = adjacency_matrix->get_size();
+        permutation_ = PermutationMatrix::create(cpu_exec, dim);
+
+        // To make it explicit.
+        inv_permutation_ = nullptr;
+        if (parameters_.construct_inverse_permutation) {
+            inv_permutation_ = PermutationMatrix::create(cpu_exec, dim);
+        }
+
+        this->generate(cpu_exec, std::move(adjacency_matrix));
+
+        // Copy back results to gpu if necessary.
+        if (is_gpu_executor) {
+            const auto gpu_exec = this->get_executor();
+            auto gpu_perm = PermutationMatrix::create(gpu_exec, dim);
+            gpu_exec->copy_from(cpu_exec.get(), dim,
+                                permutation_->get_const_permutation(),
+                                gpu_perm->get_permutation());
```

Comment:
Can this be simplified by using PermutationMatrix::copy_from(...)?

Reply:
Well, using `gpu_perm->copy_from(inv_permutation_.get())` doesn't work, throws `std::bad_typeid` at runtime. I guess that's what you meant?

Reply:
That's weird, shouldn't happen. In case you have everything setup for that, can you print a backtrace at that point? Otherwise I will take a look at it tomorrow.

Reply:
My bad, I actually executed wrote it just like that, that `inv_` must have sneaked in there. Was still a nullptr at that point. Now everything works, reads a lot nicer.